Update the tutorial to Tokio 0.2 (#352)
tokio-rs/website: content/docs/getting-started/futures.md

    parent: getting_started
---

> Let's take a closer look at futures. Tokio is built on top of `task` and `future` modules

Reviewer: Nit: should `task` and `future` be linked? `futures` was linked previously.

> found in the Rust standard library as well as the `futures` crate.

Reviewer: Nit: probably should link to `futures` here.

Reviewer: ```suggestion
found in the Rust standard library, as well as on the `futures` crate.
```

```rust,ignore
trait Future {
    /// The type of the value returned when the future completes.
    type Output;

    /// The function that will be repeatedly called to see if the future
    /// has completed or not
    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output>;
}
```

For now it's just important to know that futures have an associated type: `Output`.

> It is the type of the value that the `Future` will yield when it completes.

Reviewer: ```suggestion
This is the type of the value that the `Future` will yield when it completes.
```

> Additionally, `Future`s have one method named `poll`. We won't go into too much detail

Reviewer: Nit: I think there should be a comma here
```suggestion
Additionally, `Future`s have one method, named `poll`. We won't go into too much detail
```

about `poll` in this section since you don't need to know about `poll` to use

> futures with combinators. The only thing to be aware for now is that `poll` is

Reviewer: I might change "combinators" to "async/await" here:
```suggestion
futures with async/await. The only thing to be aware for now is that `poll` is
```

> what the runtime will call in order to see if the `Future` is complete yet or not.

Reviewer: A comparison I like to make that may be worth including is that the relationship between `Future::poll` and the `async`/`await` keywords is very similar to the relationship between `Iterator::next` and `for` loops: you don't need to be aware of the details of how `next` is implemented to iterate over a slice in a `for` loop. Futures and async/await are similar.

> If you're curious: `Poll` is an enum with values `Ready(Item)` or `Pending` which

Reviewer: Nit: should this link to https://doc.rust-lang.org/std/task/enum.Poll.html?

    /// The type of the value yielded by the stream.
    type Item;

    /// The function that will be repeatedly called to see if the stream has
    /// another value it can yield

>     fn poll_next(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>>

Reviewer: Nit: should this be 
```suggestion
    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>>
```
since the Rust 2018 idiom lints suggests that? 

I notice that the standard library definition of `Future::poll` does not use this syntax though, so maybe this should be the same?
